### PR TITLE
21 feature user api

### PR DESCRIPTION
**Pull request by albertparedandan**

### CODE REVIEW

File: server/api/controllers/baptisms/update-baptism.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        if (data != null) {
+          return exits.success(data);
+        }
+        return exits.invalid();
+      } catch (err) {
+        sails.log.error(err);
+        return exits.error(err);
+      }
+    } else {
+      throw "missingRequiredFields";
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/update-baptism.js:49]: this throw is not handled properly

File: server/api/controllers/memberships/create-membership.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      type: "string",
+    },
+    userId: {
+      type: "string",
+    },
+    classDate: {
+      type: "json",
+    },
+    classAttendance: {
+      type: "boolean",
```
Comment (albertparedandan) [inline server/api/controllers/memberships/create-membership.js:17]: how come this is type boolean if classDate is json(array of classDates). Shouldn't that mean classAttendance should be an array of boolean? or vice versa

File: server/api/controllers/memberships/update-membership.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        }).set(toUpdate);
+        if (data != null) {
+          return exits.success(data);
+        }
+        return exits.invalid();
+      } catch (err) {
+        sails.log.error(err);
+        return exits.error(err);
+      }
+    } else {
```
Comment (albertparedandan) [inline server/api/controllers/memberships/update-membership.js:48]: same comment as this is not handled properly

File: server/api/controllers/users/update-user.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+          return exits.success(data);
+        }
+        return exits.invalid();
+      } catch (err) {
+        sails.log.error(err);
+        return exits.error(err);
+      }
+    } else {
+      throw "missingRequiredFields";
+    }
```
Comment (albertparedandan) [inline server/api/controllers/users/update-user.js:50]: same comment on not being handled properly

File: server/api/models/Membership.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    classAttendance: {
+      type: "boolean",
+      defaultsTo: false,
+    },
+    recognitionDate: {
+      type: "ref",
+      columnType: "datetime",
+      defaultsTo: null,
+    },
+    recommitmentDate: {
```
Comment (albertparedandan) [inline server/api/models/Membership.js:24]: should be a JSON to keep track of the recommitment dates

File: server/api/models/user.js [hunk 6]
```diff
@@ -10,13 +10,20 @@ module.exports = {
       type: "string",
```
Comment (albertparedandan) [inline server/api/models/user.js:10]: User.js file name

File: server/api/controllers/baptisms/create-baptism.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  description: "Create a baptism record",
+
+  inputs: {
+    officialName: {
+      type: "string",
+    },
+    userId: {
+      type: "string",
+    },
+    classDate: {
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/create-baptism.js:13]: combine classDate classAttendance to just 1 object
```
{
  20210731: true, 
  yyyymmdd: boolean,
}
```

File: server/api/controllers/memberships/create-membership.js (hunk 2 above)
Comment (tamarayustian) [reply server/api/controllers/memberships/create-membership.js:17]: `classAttendance: {
"date": true
}`

File: server/api/models/Membership.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    userId: {
+      model: "User",
+      unique: true,
+    },
+    officialName: {
+      type: "String",
+    },
+    classAttendance: {
+      type: "json",
+      defaultsTo: { yyyymmdd: false },
```
Comment (albertparedandan) [inline server/api/models/Membership.js:12]: What does this do?

File: server/api/models/User.js [hunk 9]
```diff
@@ -21,13 +21,25 @@ email status until they click the link in the confirmation email.`,
       type: "string",
     },
     accessType: {
-      type: "string", //stewardship, admin, or member
+      type: "string", //stewardship, admin, alumni, signed, unsigned
```
Comment (albertparedandan) [inline server/api/models/User.js:24]: No need to action now, but can look into `isIn`

File: server/api/models/Membership.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
+  attributes: {
+    userId: {
+      model: "User",
+      unique: true,
+    },
+    officialName: {
+      type: "String",
```
Comment (albertparedandan) [inline server/api/models/Membership.js:8]: Required?

Comment (albertparedandan) [reply server/api/models/Membership.js:8]: All the below too

File: server/api/models/Baptism.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
+module.exports = {
```
Comment (albertparedandan) [inline server/api/models/Baptism.js:1]: Same comments for Memberships.js

File: server/api/controllers/memberships/create-membership.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
+module.exports = {
+  friendlyName: "Create memberships",
+
+  description: "Create a member record",
+
+  inputs: {
+    officialName: {
```
Comment (albertparedandan) [inline server/api/controllers/memberships/create-membership.js:7]: Required? All the below too

File: server/api/controllers/memberships/create-membership.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  exits: {
+    success: {
+      description: "Membership record created successfully",
+    },
+    invalid: {
+      responseType: "badRequest",
+      description: "Failed to create membership record",
+    },
+
+    duplicateRecords: {
```
Comment (albertparedandan) [inline server/api/controllers/memberships/create-membership.js:33]: Is this being used?

File: server/api/controllers/baptisms/update-baptism.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        if (data != null) {
+          return exits.success(data);
+        }
+        return exits.invalid();
+      } catch (err) {
+        sails.log.error(err);
+        return exits.error(err);
+      }
+    }
+    sails.log.error("missingRequiredFields");
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/update-baptism.js:49]: Mayb add a `sails.log.error(err)` too so that if it does happen, can easily debug

Comment (albertparedandan) [reply server/api/controllers/baptisms/update-baptism.js:49]: For the other update APIs too

File: server/api/controllers/baptisms/create-baptism.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
+module.exports = {
+  friendlyName: "Create baptisms",
+
+  description: "Create a baptism record",
+
+  inputs: {
+    officialName: {
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/create-baptism.js:7]: Required?

File: server/api/controllers/baptisms/create-baptism.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  exits: {
+    success: {
+      description: "Baptism record created successfully",
+    },
+    invalid: {
+      responseType: "badRequest",
+      description: "Failed to create baptism record",
+    },
+
+    duplicateRecords: {
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/create-baptism.js:33]: Same as the other file, is it being used?

Comment (tamarayustian) [reply server/api/controllers/baptisms/create-baptism.js:33]: not being used

File: server/api/controllers/memberships/create-membership.js (hunk 13 above)
Comment (tamarayustian) [reply server/api/controllers/memberships/create-membership.js:33]: not being used. actually it only checks for one-on-one relation right now. so it wont have duplicates i think

File: server/api/controllers/memberships/update-membership.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    invalid: {
+      description: "Failed to update membership record",
+    },
+
+    missingRequiredFields: {
+      statusCode: 409,
+      description: "Please fill in the required fields.",
+    },
+
+    invalidBaptismId: {
```
Comment (albertparedandan) [inline server/api/controllers/memberships/update-membership.js:26]: Not being used

File: server/api/controllers/baptisms/create-baptism.js [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    baptismPlace: {
+      type: "string",
+    },
+  },
+
+  exits: {
+    success: {
+      description: "Baptism record created successfully",
+    },
+    invalid: {
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/create-baptism.js:30]: Not used

File: server/api/controllers/baptisms/update-baptism.js [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    invalid: {
+      description: "Failed to update baptism record",
+    },
+
+    missingRequiredFields: {
+      statusCode: 409,
+      description: "Please fill in the required fields.",
+    },
+
+    invalidBaptismId: {
```
Comment (albertparedandan) [inline server/api/controllers/baptisms/update-baptism.js:26]: Not used

**Review by ksuhartono97 — APPROVED** [review]

**Pull request merged**